irods_specific_implementation/irods_rest_meta_api_implementation.h

```diff
@@ -0,0 +1,161 @@
+
+#include <pistache/router.h>
+#include <nlohmann/json.hpp>
+
+
+#define OP_STATUS_SUCCESS "SUCCESS"
+#define OP_STATUS_FAILED "FAILURE"
+
+namespace irods::rest
+{
+
+    namespace ixm = irods::experimental::metadata;
+
+    // this is contractually tied directly to the api implementation
+    const std::string service_name{"irods_rest_cpp_meta_server"};
+
+    class meta : public api_base
+    {
+    public:
+        meta()
+            : api_base{service_name}
+        {
+            info("Endpoint initialized.");
+        }
+
+        std::tuple<Pistache::Http::Code, std::string>
+        operator()(const Pistache::Rest::Request& _request,
+                   Pistache::Http::ResponseWriter& _response)
+        {
+            try {
+                auto _attrib       = _request.query().get("attrib").get();
+                auto _value        = _request.query().get("value").get();
+                auto _unit         = _request.query().get("unit").get();
+                auto _entity_type  = _request.query().get("entity_type").get();
+                auto _sub_cmd      = _request.query().get("sub_cmd").get();
+                auto _target       = _request.query().get("target").get();
+
+                // unsure what the hint parameter achieves here
+                auto conn = get_connection(_request.headers().getRaw("authorization").value());
+
+                nlohmann::json json_obj;
+
+                // construct avu struct
+                ixm::avu triple {
+                    _attrib,
+                    _value,
+                    _unit
+                };
+                // i think the target parameter is constructed at a lower level of abstraction
+                // get entity type
+                auto et = ixm::to_entity_type("-" + _entity_type);
+
+                if (_sub_cmd == "add"){
+                        ixm::add(
+                            *conn(),
+                            triple,
+                            et,
+                            _target
+                        );
+                        json_obj["op"] = "add";
+                        json_obj["op_status"] = OP_STATUS_SUCCESS;
+                        json_obj["avu"] = {
+                            { "attrib", _attrib },
+                            { "value", _value },
+                            { "unit", _unit }
+                        };
+                        json_obj["target"] = _target;
+                        json_obj["entity_type"] = _entity_type;
+                        return std::make_tuple(
+                            Pistache::Http::Code::Ok,
+                            json_obj.dump(
+                                -1,
+                                ' ',
+                                false,
+                                nlohmann::json::error_handler_t::ignore
+                            )
+                        );
+                } else if (_sub_cmd == "rm") {
+                        ixm::remove(
```

Comment:
Do these functions throw?
How are errors reported/handled here?

Reply:
If I understand some discussion from the NLohmann's GitHub page, these functions shouldn't throw (the dumps definitely won't). But it's only true that we could look for a request header -- if we get bad text or something, _and_ we know the text encoding, we can just translate to UTF-8, for example. Will make this more robust.

The filesystem operation errors are caught by the "make_error_response" catches at the bottom.

Reply:
Those could throw various exceptions if the operation is invalid. The catch block at the bottom log those errors and send an HTTP response with a description of the error.